internal/store/cache_stores_snapshot.go

```diff
@@ -21,6 +27,68 @@ func (c CacheStores) TakeSnapshot() (CacheStores, error) {
 	return snapshot, err
 }
 
+// TakeSnapshotIfChanged takes a snapshot of the CacheStores if the hash of the current state
+// differs from the hash of the previous snapshot supplied as an argument (to make and initial
+// just pass empty string). When error is not nil discard all other return values.
+// When newHash is empty it means that the snapshot hasn't been taken - returned snapshot is
+// meaningless. This is a situation when hash of the current state is the same as the hash of
+// the previous snapshot supplied as an argument.
+func (c CacheStores) TakeSnapshotIfChanged(previousSnapshotHash string) (
+	snapshot CacheStores,
+	newHash string,
+	err error,
+) {
+	// Initialize all variables that don't need to be guarded by a lock.
+	snapshot = NewCacheStores()
+	listOfStores := c.listAllStores()
+	accessor := meta.NewAccessor()
+	hashCalculator := sha256.New()
+	var capturedErr error
+
+	c.l.RLock()
+	defer c.l.RUnlock()
+
+	// Compute the hash of the current store.
+	for _, store := range listOfStores {
+		// Underlying store is implemented a thread-safe map so for method List() it doesn't maintain order of items.
+		// To successfully calculate hash we need to sort the items.
```

Comment:
```suggestion

	c.l.RLock()
	defer c.l.RUnlock()

	// Compute the hash of the current store.
	for _, store := range listOfStores {
		// Underlying store is implemented a thread-safe map so for method List() it doesn't maintain order of items.
		// To successfully calculate hash we need to sort the items.
		var capturedErr error
```
Let's define `capturedErr` as close as possible to its usage (`lo.Map` callback func).